elyra/pipeline/component_catalog.py

```diff
@@ -72,90 +70,165 @@ class ComponentCatalog(LoggingConfigurable):
                               extensions=[".r"],
                               categories=[_generic_category_label])}
 
-    ttl_default = 300
-    cache_ttl_env = 'ELYRA_COMPONENT_CATALOG_CACHE_TTL'
-    cache_ttl = Integer(ttl_default,
-                        help="Time-to-live (in seconds) for Component Catalog cache entries. "
-                             "(ELYRA_COMPONENT_CATALOG_CACHE_TTL env var)").tag(config=True)
-
-    @default('cache_ttl')
-    def cache_ttl_default(self):
-        ttl = ComponentCatalog.ttl_default
-        try:
-            ttl = int(os.getenv(self.cache_ttl_env, ttl))
-        except ValueError:
-            pass
-        return ttl
-
-    def __init__(self,
-                 parser: ComponentParser,
-                 caching_enabled: bool = True,
-                 **kwargs):
+    def __init__(self, **kwargs):
         super().__init__(**kwargs)
-        self._parser = parser
 
-        # Initialize the cache
-        self.caching_enabled = caching_enabled
-        if self.caching_enabled:
-            self.log.debug(f"ComponentCatalog cache TTL: {self.cache_ttl}")
-            self.update_cache()
+        # Indicate whether this is a test environment. This is relevant in
+        # update_cache_for_catalog() - we want to wait for the cache updater
+        # thread to finish in test envs
+        self._for_test = bool(kwargs.get('for_test'))
+
+        self._cache_lock = threading.Lock()
+        if not self._component_cache:
+            self._build_cache()
 
-    def get_all_components(self) -> List[Component]:
+    def _build_cache(self):
         """
-        Retrieve all components; use the component catalog cache if enabled
+        Reads through all catalog Metadata instances to build the initial component cache
         """
-        if self.caching_enabled:
-            if self._is_cache_expired():
-                self.update_cache()
-            return list(self._cached_components.values())
 
-        return list(self._read_component_catalogs().values())
+        all_catalogs = MetadataManager(schemaspace=ComponentCatalogs.COMPONENT_CATALOGS_SCHEMASPACE_ID).get_all()
+        for catalog in all_catalogs:
+            self.update_cache_for_catalog(catalog)
 
-    def get_component(self, component_id: str) -> Optional[Component]:
+    def update_cache_for_catalog(self, catalog: Metadata, operation: Optional[str] = None):
```

Comment:
Hmm I ran into 2 issues when trying to change this

1. since the call to `MetadataManager(...).get_all()` in line 91 returns `Metadata`, we get typehint errors when calling `update_cache_for_catalog` -- "Expected type 'ComponentCatalogMetadata', got 'Metadata' instead" 
2. there's a circular import reference (importing `ComponentCatalogMetadata` in `component_catalog` and importing `ComponentCatalog` in `component_metadata`) that we'd have to work around

Thoughts on the best way forward?

Reply:
Regarding 1, this would be a bug.  I can see a raw `Metadata` instance returned in error situations, but the `from_dict()` method should leverage any existing `metadata_class_name` value.  Are you able to step through `get_all()` to see why these aren't `ComponentCatalogMetadata` instances?  Since the cache update is obviously getting triggered, we know the schema is pointing at the right class name.

For 2, these kinds of things can be difficult.  Since `ComponentCatalogMetadata` is just a metadata representation of a `ComponentCatalog`, one could argue that its definition also be included adjacent to `ComponentCatalog`'s definition (i.e., in the same file).  This may be worth a try, although I may not resolve the circular reference.  There are other approaches we could take (I believe), but let's try this first since it makes sense from a definition-location standpoint.

Reply:
In order to avoid the need to update the example catalog schema files, we'll have to address the circular dependency issue by importing the entire `component_catalog` module in `component_metadata.py` and calling the `ComponentCache` instance using `component_catalog.ComponentCache.instance()`